ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/PlanModifierForASTConv.java

```diff
@@ -230,6 +225,47 @@ private static void convertOpTree(RelNode rel, RelNode parent) {
     }
   }
 
+  /**
+   * A handler that detects self-joins in the plan and rewrites them to resolve ambiguous aliases.
+   * The handler traverses the plan and collects aliases of tables it encounters for each join branch.
+   * When the same alias occurs in both branches of a join, it introduces a derive table (Project)
+   * over the left branch to break the ambiguity.
+   */
+  private static class SelfJoinHandler extends RelHomogeneousShuttle {
+    private final Set<String> aliases = new HashSet<>();
+
+    @Override
+    public RelNode visit(final RelNode rel) {
+      if (rel instanceof Join join) {
```

Comment:
Can `HiveRelShuttleImpl` be used instead of `RelHomogeneousShuttle` to get rid of `rel instanceof <Something>`?